File: CloudCoreCom/Server/src/main/java/de/cloud/core/com/server/handlers/NetworkHandler.java
package de.cloud.core.com.server.handlers;

import de.cloud.core.common.packet.Packet;
import de.cloud.core.com.server.Server;
import io.netty.channel.Channel;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.SimpleChannelInboundHandler;

public class NetworkHandler extends SimpleChannelInboundHandler<Packet> {

    private Channel channel;
    private final Server server;

    public NetworkHandler(Server server) {
        this.server = server;
    }

    @Override
    public void channelActive(ChannelHandlerContext ctx) throws Exception {
        this.channel = ctx.channel();
    }


    @Override
    protected void channelRead0(ChannelHandlerContext ctx, Packet packet) throws Exception {

    }

    @Override
    public void channelUnregistered(ChannelHandlerContext ctx) throws Exception {
        super.channelUnregistered(ctx);
        this.server.disconnect(ctx.channel());
    }
}

File: CloudCoreCom/Server/src/main/java/de/cloud/core/com/server/handlers/NetworkHandler.kt
package de.cloud.core.com.server.handlers

import de.cloud.core.api.ClientType
import de.cloud.core.com.server.Server
import de.cloud.core.common.packet.*
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.SimpleChannelInboundHandler

class NetworkHandler(val server: Server) : SimpleChannelInboundHandler<Packet>() {

    override fun channelRead0(ctx: ChannelHandlerContext, msg: Packet) {
        when (msg) {
            is PacketInAuthentication -> server.connectionManager.acceptAuthPacket(ctx.channel(), msg)
            is PacketInLink -> server.connectionManager.send(
                ClientType.MINECRAFT,
                PacketOutRequestLink(msg.ip, msg.dbId, msg.username, msg.type)
            )
            is PacketInRequestLink -> server.connectionManager.send(msg.type, PacketOutLink(msg.dbId, msg.stats))
        }
    }

    override fun channelUnregistered(ctx: ChannelHandlerContext) {
        super.channelUnregistered(ctx)
        server.disconnect(ctx.channel())
    }
}